src/main/index.js

```diff
@@ -77,5 +77,14 @@ app.on('activate', () => {
   }
 });
 
+// Emitted when calling app.quit(), or when window-all-closed is not defined.
+app.on('will-quit', (event) => {
+  // If there is a save pending, stop exiting and wait before the save is complete.
+  if (saveStateHandler.currentPromise) {
+    event.preventDefault();
+    saveStateHandler.currentPromise.then(() => app.quit());
```

Comment:
Wouldn't be safer to invert the 2 lines above just in case the saveStateHandler.currentPromise end after  `event.preventDefault();` but before `saveStateHandler.currentPromise.then(() => app.quit());`

Reply:
It's single threaded, there are no concurrency issues. In any case, when a promise fulfills, you can call `.then()` a hundred times on it, it will just return the result given to `resolve()`.

Reply:
`event.preventDefault` is synchronous anyway

Reply:
Please correct me where i'm wrong :

It's not a concurrency issue but a "non blocking" issue. There is 1 thread but, with asynchronous code, you can't know in which order it will run the code lines.
Yes, you can call a .then to an already resolved promise and it works, but in our case. The promise reference is deleted in the first .then()

So the bad scenario is the following : 
```javascript
// saveStateHandler.currentPromise is defined
if (saveStateHandler.currentPromise) {
  event.preventDefault();
// The saveStateHandler.currentPromise is resolved and the main process decide to run the first .then of it

 .then(() => {
+        this.currentPromise = null;
+   // The promise's reference is deleted
+        if (this.pendingSave) {
+          const tree = this.pendingSave;
+          this.pendingSave = null;
+          return this.onSaveState(tree);
+        }
+
+        return null;
+      });

// The process go back to the first peace of code  : 

saveStateHandler.currentPromise.then(() => app.quit());
// a Uncaught TypeError: Cannot read property 'then' of null is thrown :-S

````

Reply:
I see what you mean, and I actually forgot to set the currentPromise to the return value in the `.then()` (recursive). Although, that would not crash as you say because the `if` condition and the `.then` are in the same synchronous block, so we can be sure that they will be executed in the right order.

Reply:
Actually I didn't forget to set currentPromise, it's done by the recursivity.